add report for auditing contents of technical metadata database

jmartin-sul opened the pull request:
## Why was this change made? 🤔

to run the audit described in https://github.com/sul-dlss/technical-metadata-service/issues/515 (will close that issue once the report has actually been run)

## How was this change tested? 🤨

on QA.  once the related puppet changes were applied and the supporting shared_configs updated, i `scp`ed the report code over.  then tested, fixed, refactored for usability, tested, fixed, repeated, until this felt like it'd be likely to run through on prod (or produce helpful logging if it didn't).

i did one full run on QA with this code (448 druids with techMD inconsistencies were found), and also ran both the test report and the single druid method.

example test report stdout (same stdout as full report would be, just limited to 20 druids audited):
```
$ bin/rails r -e p 'AuditTechnicalMetadataFileList.test_report'
druid:bk509qs2277: found technical-metadata-service database; inconsistencies with v11 cocina: {"missing_filenames"=>["asawa.jp2", "maps-africa.jp2", "moore.jp2", "hay.jp2"]}
druid:ww089jm4506: not found in technical-metadata-service database
druid:wy216td0310: found technical-metadata-service database; inconsistencies with v1 cocina: {"missing_filenames"=>["example.jp2"]}
druid:xb970cg0663: found technical-metadata-service database; inconsistencies with v1 cocina: {"missing_filenames"=>["example.jp2"]}
druid:gv185zf2539: not found in technical-metadata-service database
```

progress logging worked, as did the other info and debug logging.

example running on a single druid:
```
$ bin/rails r -e p 'AuditTechnicalMetadataFileList.audit_one_druid(druid: "druid:ww089jm4506")'
druid:ww089jm4506: not found in technical-metadata-service database
```

⚡ ⚠ If this change has cross service impact, including data writes to shared file systems, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

jmartin-sul commented:
interestingly, the QA audit turned up some unexpected files in techMD DB, and some instances where file names matched for a druid but md5 differed.  i was only expecting missing files, but i wasn't sure enough of my understanding of the original problem for that to be a strong expectation.  either way, glad i did had the techMD endpoint return all 3 categories of problem.

jmartin-sul commented on app/reports/audit_technical_metadata_file_list.rb:
```diff
@@ -0,0 +1,185 @@
+  class << self
+    # @param [Array] an array of arrays.  each element in the top level list is an array containing
+    #   druid, version, and cocina-models file hashes grouped by resource (plucked from Dro)
+    def process_dro_row(dro_row, techmd_connection, logger)
```
perhaps this should've been a service class.  but i think the constructor would take these 3 params, which would only save one param among the other class methods (`dro_row` to `.druid_version_files`).  so i'm not sure much concision would be gained from a service class.  and this feels like an ok amount of reusability (`.process_drow_row`) and boxing of complexity into private helper methods.  so i'm inclined to leave this factored as-is, since i'm not sure it's worth spending more time cleaning this up (though happy to take suggestions if i'm missing something obvious).

justinlittman commented on config/settings.yml:
```diff
@@ -74,6 +74,10 @@ purl_fetcher:
   url: 'https://purl-fetcher.example.edu'
   token: 'fake-token'
 
+technical_metadata:
```
Maybe a note why this config is here?

jmartin-sul replied:
good call, will do

justinlittman commented on app/reports/audit_technical_metadata_file_list.rb:
```diff
@@ -0,0 +1,185 @@
+    @techmd_connection ||= Faraday.new(Settings.technical_metadata.url) do |builder|
+      builder.request :retry, retry_options
+      builder.use Faraday::Request::UrlEncoded
+      # builder.use Faraday::Response::RaiseError # raise exceptions on 40x, 50x responses
```
Cruft

jmartin-sul replied:
yes, good catch

justinlittman commented on app/reports/audit_technical_metadata_file_list.rb:
```diff
@@ -0,0 +1,185 @@
+  def retry_options
+    {
+      max: 3,
+      methods: %i[delete get head options put post]
```
Just `get`?

jmartin-sul replied:
in this case `post` is the only type of request we make, and is actually retriable, because it's just a read with no side-effects.  should've been a `get` semantically, but the file list we're sending could overrun the traditional max size for URL params in a `GET`, hence the `POST` usage.

i could make the list just `post`.  was on the fence about that, felt a little weird.  either way, seems like a comment would be good since this is an unusual config?

justinlittman reviewed (approved):
Approved with comments.

jmartin-sul commented:
added a comment to `settings.yml`, got rid of the Faraday config cruft in `#techmd_connection`, added a comment to the confusing `methods` list in `#retry_options`.  squashed that into the existing commit, but here's the diff from before i pushed for easy review:

```diff
% git diff origin/audit-techmd HEAD
diff --git a/app/reports/audit_technical_metadata_file_list.rb b/app/reports/audit_technical_metadata_file_list.rb
index aa94cc72..483620d5 100644
--- a/app/reports/audit_technical_metadata_file_list.rb
+++ b/app/reports/audit_technical_metadata_file_list.rb
@@ -154,6 +154,9 @@ class AuditTechnicalMetadataFileList
   def retry_options
     {
       max: 3,
+      # The default retriable/idempotent HTTP methods, plus POST. As of May 2024, POST is the only
+      # HTTP method this class actually uses (for a read-only operation, but the file list parameter
+      # might exceed the traditional max for total URL query param size, hence POST instead of GET).
       methods: %i[delete get head options put post]
     }
   end
@@ -162,7 +165,6 @@ class AuditTechnicalMetadataFileList
     @techmd_connection ||= Faraday.new(Settings.technical_metadata.url) do |builder|
       builder.request :retry, retry_options
       builder.use Faraday::Request::UrlEncoded
-      # builder.use Faraday::Response::RaiseError # raise exceptions on 40x, 50x responses
       builder.adapter Faraday.default_adapter
       builder.headers[:user_agent] = "dor-services-app #{Socket.gethostname}"
       builder.headers['Authorization'] = "Bearer #{Settings.technical_metadata.token}"
diff --git a/config/settings.yml b/config/settings.yml
index e2593ca8..a3fc94aa 100644
--- a/config/settings.yml
+++ b/config/settings.yml
@@ -74,6 +74,9 @@ purl_fetcher:
   url: 'https://purl-fetcher.example.edu'
   token: 'fake-token'
 
+# As of may 2024, this is just here to support auditing contents of
+# technical-metadata-service. If the report that uses it (AuditTechnicalMetadataFileList)
+# is removed, see whether this section can be removed.
 technical_metadata:
   url: 'https://techmd-svc.example.edu'
   token: 'secret-token'
```

merging since approved and i think i addressed the comments, but happy to revisit the `retry_options` thing or whatever else in a follow up PR if you'd like.
